### ISSUE TITLE
Bug: [strict-boolean-expression] using string literal union bypasses allowString: false

### ISSUE THREAD

**Issue by wolverineks** [typescript-eslint/typescript-eslint#6624]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.1-rc&sourceType=module&code=MYewdgzgLgBARgQwE4wLwwBQDMQgFwwBEChMAPkaRQK5gAmAplgJZgN0CUaAfDAN4AoGDGZZMAQhwgug4cNCQQAGwYA6JSADm2XByEwAvgINA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs+SAIZN6AD2LQUySnyaoMAbXDYciaND7RIKiAF0VAXxCGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
const bar = (foo: "a" | "" | undefined) => {
  if (!foo) {
    console.log(foo)
  }
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/strict-boolean-expressions": [
      "error"
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

i expect line 2 to error with
```
Unexpected nullable string value in conditional. Please handle the nullish/empty cases explicitly.
```

### Actual Result

no error

**Comment by bradzacher** [comment]
The bug here seems to be if you have multiple literal types of the same type AND one of them is the falsey literal type, then we don't properly check them.

The same thing happens for numbers (eg if the type was `0 | 1 | undefined`, it wouldn't get flagged).
